target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mf289f.dts

```diff
@@ -0,0 +1,433 @@
+        };
+
+};
+
+&blsp1_uart1 {
+        pinctrl-0 = <&serial_pins>;
+        pinctrl-names = "default";
+        status = "okay";
+};
+
+&cryptobam {
+        status = "okay";
+};
+
+&gmac0 {
+        nvmem-cell-names = "mac-address";
+        nvmem-cells = <&macaddr_config_0>;
+};
+
+&gmac1 {
+        nvmem-cell-names = "mac-address";
+        nvmem-cells = <&macaddr_config_0>;
+        mac-address-increment = <1>;
+};
+
+
+&qpic_bam {
+        status = "okay";
+};
+
+&tlmm {
+        i2c_0_pins: i2c_0_pinmux {
+                mux {
+                        pins = "gpio20", "gpio21";
+                        function = "blsp_i2c0";
+                        bias-disable;
+                };
+        };
+
+        mdio_pins: mdio_pinmux {
+                mux_1 {
+                        pins = "gpio6";
+                        function = "mdio";
+                        bias-pull-up;
+                };
+
+                mux_2 {
+                        pins = "gpio7";
+                        function = "mdc";
+                        bias-pull-up;
+                };
+        };
+
+        serial_pins: serial_pinmux {
+                mux {
+                        pins = "gpio16", "gpio17";
+                        function = "blsp_uart0";
+                        bias-disable;
+                };
+        };
+
+        spi_0_pins: spi_0_pinmux {
+                pinmux {
+                        function = "blsp_spi0";
+                        pins = "gpio13", "gpio14", "gpio15";
+                        drive-strength = <12>;
+                        bias-disable;
+                };
+
+                pinmux_cs {
+                        function = "gpio";
+                        pins = "gpio12", "gpio54";
+                        drive-strength = <2>;
+                        bias-disable;
+                        output-high;
+                };
+        };
+};
+
+&usb2_hs_phy {
+        status = "okay";
+};
+
+&usb3_ss_phy {
+        status = "okay";
+};
+
+&usb3_hs_phy {
+        status = "okay";
+};
+
+&wifi0 {
+        status = "okay";
+        nvmem-cell-names = "pre-calibration", "mac-address";
+        nvmem-cells = <&precal_art_1000>, <&macaddr_config_0>;
+        mac-address-increment = <2>;
+        qcom,ath10k-calibration-variant = "zte,mf289f";
+};
+
+&wifi1 {
+        status = "okay";
+        nvmem-cell-names = "pre-calibration", "mac-address";
+        nvmem-cells = <&precal_art_5000>, <&macaddr_config_0>;
+        mac-address-increment = <2>;
+        qcom,ath10k-calibration-variant = "zte,mf289f";
+};
+
+&pcie0 {
+        status = "okay";
+        perst-gpio = <&tlmm 38 GPIO_ACTIVE_LOW>;
+        wake-gpio = <&tlmm 40 GPIO_ACTIVE_LOW>;
+        clkreq-gpio = <&tlmm 39 GPIO_ACTIVE_LOW>;
+ 
+        bridge@0,0 {
+                 reg = <0x00000000 0 0 0 0>;
+                 #address-cells = <3>;
+                 #size-cells = <2>;
+                 ranges;
+ 
+                 wifi2: wifi@1,0 {

```

Comment:
It might be worth commenting here, that some variants use "external" PCIe Wi-Fi chip, while others use built-in radio. Is it possible to detect, that 5GHz path on the IPQ4019 side is not used in such case?

Reply:
I've tested on both variant, what is change:

AT1: there is a 5ghz radio on IPQ4019 that is using board-id 31, not present on board-2.bin file, so this will not loaded for "missing firmware", in that case the dts will found the 5ghz on PCIe and load firmware from QCA9984

AT2: radios are both on IPQ4019 and they are using board-id 20 for 2.4ghz and 21 for 5ghz

if there is a best approch let me know

Reply:
Please correct me if I'm wrong, but if final board-2.bin shipped in image contains all the IDs, then system will actually enable both 5GHz radios, even the non-functional one?

From user perspective, if there is one image that can support both variants, it's better - and personally I would go for it - but errors in kernel log for AT1 variant may be confusing, when the system fails to enable the radio due to missing board ID.

Reply:
No,

Until board-2 image is not updated from ath10k repository, the AT2 variant will match all Wi-Fi configuration of MF286D (same board od), while on AT1 there will be only 2.4Ghz match (board  id 21), while 5Ghz will get the PCIE one because the IPQ4019 board id 31 was not added to board-2

Reply:
I'm not sure if I understand that correctly. Do you have kernel logs from both variants available? Maybe this would clarify the device behaviour to me.
I presume you installed the same build for both variants for testing.

Reply:
this is a bootlog from AT1:

```
[   15.468317] ath10k 5.15 driver, optimized for CT firmware, probing pci device: 0x46.
[   15.469271] ath10k_pci 0000:01:00.0: enabling device (0140 -> 0142)
[   15.475788] ath10k_pci 0000:01:00.0: pci irq msi oper_irq_mode 2 irq_mode 0 reset_mode 0
[   17.317088] ath10k_pci 0000:01:00.0: qca9984/qca9994 hw1.0 target 0x01000000 chip_id 0x00000000 sub 168c:cafe
[   17.317153] ath10k_pci 0000:01:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   17.331875] ath10k_pci 0000:01:00.0: firmware ver 10.4b-ct-9984-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 7ea63dc5
[   19.622494] ath10k_pci 0000:01:00.0: Loading BDF type 0
[   19.624614] ath10k_pci 0000:01:00.0: failed to fetch board data for bus=pci,vendor=168c,device=0046,subsystem-vendor=168c,subsystem-device=cafe,variant=zte,mf289f from ath10k/QCA9984/hw1.0/board-2.bin
[   19.829859] ath10k_pci 0000:01:00.0: failed to fetch board-2.bin or board.bin from ath10k/QCA9984/hw1.0
[   19.829947] ath10k_pci 0000:01:00.0: failed to fetch board file: -12
[   19.838456] ath10k_pci 0000:01:00.0: could not probe fw (-12)
[   21.544843] ath10k_ahb a000000.wifi: qca4019 hw1.0 target 0x01000000 chip_id 0x003b00ff sub 0000:0000
[   21.544896] ath10k_ahb a000000.wifi: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   21.556767] ath10k_ahb a000000.wifi: firmware ver 10.4b-ct-4019-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 6b2b5c5b
[   21.616848] ath10k_ahb a000000.wifi: Loading BDF type 0
[   21.711148] ath10k_ahb a000000.wifi: board_file api 2 bmi_id 0:20 crc32 de0d18b2
[   23.007667] ath10k_ahb a000000.wifi: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
[   23.007717] ath10k_ahb a000000.wifi: msdu-desc: 2500  skid: 32
[   23.056718] ath10k_ahb a000000.wifi: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
[   23.058424] ath10k_ahb a000000.wifi: wmi print 'free: 53252 iram: 13432 sram: 35752'
[   23.198540] ath10k_ahb a000000.wifi: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-nvmem max-sta 32 raw 0 hwcrypto 1
[   24.950582] ath10k_ahb a800000.wifi: qca4019 hw1.0 target 0x01000000 chip_id 0x003b00ff sub 0000:0000
[   24.950650] ath10k_ahb a800000.wifi: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   24.962546] ath10k_ahb a800000.wifi: firmware ver 10.4b-ct-4019-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 6b2b5c5b
[   25.022303] ath10k_ahb a800000.wifi: Loading BDF type 0
[   25.022496] ath10k_ahb a800000.wifi: failed to fetch board data for bus=ahb,bmi-chip-id=0,bmi-board-id=31,variant=zte,mf289f from ath10k/QCA4019/hw1.0/board-2.bin
[   25.215865] ath10k_ahb a800000.wifi: failed to fetch board-2.bin or board.bin from ath10k/QCA4019/hw1.0
[   25.215938] ath10k_ahb a800000.wifi: failed to fetch board file: -12
[   25.224278] ath10k_ahb a800000.wifi: could not probe fw (-12)
```

as you can see there is  **ath10k_pci**  that is referred to the QCA9984 (only avaliable on AT1 rev), and a secong 5Ghz **ath10k_ahb a800000.wifi** that have board-id=31 on AT1, while on AT2 is board-id=21 like MF286D.

So an AT2 version will no display these logs because QCA9984 is not present (so ath10k_pci will never probed) and the error for board-id=31 will not displayed because the IPQ4019 5Ghz radio on that model use board-id=21

hope is more clear now :)

Reply:
OK, now I get it, I missed the part that a0000000.wifi has different board IDs depending on revision.
I would add comments in the device tree in the nodes of respective radios, that they're used only in AT1 or in AT2 variant.

Reply:
I'd prefer it in single line, and just over wifi@1,0 node. and a similar comment over `&wifi1` node.

Reply:
done, let me know so I can mark as resolved :)

Reply:
OK, LGTM.